project1/server/agent.cpp

```diff
@@ -4,231 +4,415 @@
-}
-
-Agent::ClientState Agent::HandleRequest(
-    const ftp_messages::QuitRequest& request) {
-  std::cout << "Got a QUIT request. Exiting." << std::endl;
-  // Indicate that we are finished with this client.
-  return ClientState::DISCONNECTED;
-}
+    using file_handler::ThreadSafeFileHandler;
+    using ftp_messages::Request;
+    using ftp_messages::Response;
+
+    namespace {
+
+        /// Generic empty response.
+        const Response kEmptyResponse{};
+
+    }  // namespace
+
+    Agent::Agent(int client_fd, std::unique_ptr<ThreadSafeFileHandler> file_handler,
+                 std::shared_ptr<server_tasks::CommandIDs> active_commands)
+            : client_fd_(client_fd), active_commands_(active_commands), file_handler_(std::move(file_handler)) {}
+
+    Agent::Agent(int client_fd,
+                 std::shared_ptr<server_tasks::CommandIDs> active_commands) :
+            client_fd_(client_fd), active_commands_(active_commands) {}
+
+    Agent::~Agent() {
+        // Close the socket.
+        close(client_fd_);
+    }
+
+    bool Agent::Handle() {
+        ClientState client_state = ClientState::ACTIVE;
+
+
+        while (client_state == ClientState::ACTIVE) {
+            // Get the next message from the socket.
+            Request message;
+            client_state = ReadNextMessage(&message);
+
+            if (client_state == ClientState::ACTIVE) {
+                // Handle the message.
+                client_state = DispatchMessage(message);
+            }
+        }
+
+        return client_state != ClientState::ERROR;
+    }
+
+    Agent::ClientState Agent::ReadFileContents(ftp_messages::FileContents *fc, uint16_t command_id) {
+
+        while (!fc_parser_.HasCompleteMessage()) {
+            incoming_message_buffer_.resize(kClientBufferSize);
+
+            uint8_t buf[1000];
+
+            // Read 1000 bytes from the socket
+            const auto bytes_read =
+                    recv(client_fd_, buf, 1000, 0);
+
+            if (bytes_read < 0) {
+                // Failed to read anything.
+                perror("Failed to read from client socket");
+                fc_parser_.ResetParser();
+                return ClientState::ERROR;
+            } else if (bytes_read == 0) {
+                // Client has disconnected nicely.
+                std::cout << "Detected client disconnect." << std::endl;
+                return ClientState::DISCONNECTED;
+            }
+
+            bool terminated = !active_commands_->Contains(command_id);
+
+            // only continue if this put request is an active command
+            if (!terminated) {
+
+                // place data into the incoming message buffer
+                for (int i = 0; i < 1000; i++) {
+                    incoming_message_buffer_.push_back(buf[i]);
+                }
+
+            } else {
+                std::cout << "Command #" << command_id << " successfully terminated." << std::endl;
```

Comment:
Hmm. In this case, you probably just want to break out of the loop. Otherwise, more likely than not, `HasCompleteMessage` will continue to return false, and the termination won't actually happen.